ddtrace/ext/provider.py

```diff
@@ -0,0 +1,90 @@
+"""
+common CI providers
+"""
+
+import os
+import re
+
+from ddtrace.vendor import attr
+
+from .ci.services import (
+    appveyor,
+    azure_pipelines,
+    bitbucket,
+    buildkite,
+    circle_ci,
+    github_actions,
+    gitlab,
+    jenkins,
+    teamcity,
+    travis,
+)
+from . import ci, git
+from ..internal.logger import get_logger
+
+log = get_logger(__name__)
+
+_PROVIDER_METADATA_KEY = "ddtrace.ext.provider"
+
+
+def field(name, *args, **kwargs):
+    kwargs.setdefault("default", None)
+    kwargs.setdefault("metadata", {})
+    kwargs["metadata"][_PROVIDER_METADATA_KEY] = name
+    return attr.ib(*args, **kwargs)
+
+
+_RE_BRANCH_PREFIX = re.compile(r"^refs/(heads/)?")
+
+
+@attr.s(kw_only=True, eq=False, order=False, slots=True, frozen=True)
+class Provider(object):
```

Comment:
I will remove the `match` function as it looks like we will be able to find a unique env variable to match right CI provider, but I would prefer to keep the rest. It makes it easier to change tag names in a single place and perform validation and clean up in a single place.

Reply:
Hmm yeah I'm still not seeing the benefits that we get to all this machinery. All we're ultimately doing is

```
Provider.from_env().astags()
```

It seems overly complex that we need to have all these modules and imports just to read some environment variables. 

> It makes it easier to change tag names in a single place and perform validation and clean up in a single place

This would also be true of the rough solution I proposed:

```
def get_ci_tags():
    tags = {}
    if "TRAVIS" in os.environ:
        tags = {...}
    elif "AZURE" in os.environ:
        tags = {...}
    
    # Validate tags
    # ...
    return tags
```
it's not as pretty, but I'd argue that it's more understandable.

Reply:
I have refactor it to use same pattern as in `dd-trace-rb`. Let me know what you think about it.